### ISSUE TITLE
TLKEventListener issue

### ISSUE THREAD

**Issue by bdwilton** [LaKraven/LKSL#104]
I have an intermittent issue where I trigger a TLKEvent from within a OnPaint routine (so should be the main thread) and it doesn't seem to fire to the TLKEventListener e.g. TAnEvent.Create(FOffEvent).ScheduleQueue(0.1);
I've tried Queue as well with no luck, a larger delay does seem to keep it running longer, not sure why that is.
When it locks up I can get things running again by firing another TAnEvent by pressing a TButton.  So when I fire the new event I can see the stuck one turn up first followed by the just fired event. The TLKEventThread is owned by TForm and is just a win32 program.
I guess a test app would help, will see what I can do when I get time.

**Comment by LaKraven** [comment]
Can you please paste your implementation?

**Comment by bdwilton** [comment]
I've created a test app (in-between weekend kids taxi duties) that creates this problem where the event isn't delivered.  Not sure how one adds a test app on github so download is http://wiltonsoftware.com/download/EventBug.zip

Just run the application for a while and it will stop changing the PaintBox colors at some point.  You can then press the Manual Event button which sends a manual event and triggers things to start again.

I hacked the test code together, it seems to have an issue on exiting and destroying the LKSL.  I didn't look into what I did wrong there, my main app doesn't have this issue so must be something I removed somewhere.

**Comment by LaKraven** [comment]
Having gone through your demo, I can tell you safely that your issue isn't with the Event Engine, rather with the way you're drawing your frames.

You're attempting to *push* the frame onto the canvas, rather than *invalidating* the canvas and having *it* take care of rendering itself using the `OnPaint` method.

I encountered this same issue when developing the early prototypes for the AGE (Another Game Engine) project.

If you apply this same code to a slightly different context (such as, updating the captions of 4 respective `TLabel` controls), you'll notice that it does not freeze up.

**Comment by LaKraven** [comment]
I'll post a counter-example when I get home today... showing how I use the LKSL to produce the Render Loop for the AGE project.

You would have the same issue using a simple `TThread` and trying to update the frames as you're doing in that test project. It took me *weeks* to figure out what I was doing and how to *properly* flip the frames. I'll save you that annoying learning curve :)

**Comment by bdwilton** [comment]
Ok interesting, I'm not surpised as FM has been a pain in the proverbial so many times now!  The number of times what I thought was going to be simple that has taken hours of head scratching.

I had tried InvalidateRect(<paintbox>ClipRect); but it didn't seem to make a difference.  I had used that previously as under Android I think it wasn't drawing sometimes but the Sychronize(...repaint) seemed to get that working in this example.

I just tried InvalidateRect() and it has locked up a couple of times now so maybe its some other aspect of how I'm doing things.

**Comment by bdwilton** [comment]
If I quite a difference times I also get a pointer violation in TCriticalSection sometimes, stack trace below.  Works most of the time but now and then, hadn't noticed this previously but since I updated to a later version.

:00694F70 System::Syncobjs::TCriticalSection::Acquire(Self=????)
LKSL.Common.Types.TLKInterfacedObject.AcquireReadLock
LKSL.Events.Main.{LKSL.Generics.Collections}TLKListBase<LKSL.Events.Main.TLKEvent>.GetCount
LKSL.Events.Main.TLKEventContainer.ProcessStack(0.000432299995736685,49961.3203772)
LKSL.Events.Main.TLKEventContainer.ProcessEvents(0.000432299995736685,49961.3203772)
LKSL.Events.Main.TLKEventThread.PreTick(0.000432299995736685,49961.3203772)
LKSL.Threads.Main.TLKThread.Execute

**Comment by LaKraven** [comment]
Yes, when a synchronized event callback is being called, and you attempt to terminate the program, things go badly wrong.

I'm working on a special solution for "cleanly synchronized Event Listeners" to resolve that very issue.

**Comment by bdwilton** [comment]
I have re-gigged the code as the example previously wasn't great.  Its now a single TPaintBox where I draw four color rectangles.  Demo http://wiltonsoftware.com/download/EventBug2.zip

Now if you run the application and wait for it to stop.

When stopped add a break point on line 36 in DoEvent.pas, this will give the global event UID of the next event that is triggered.  Also add a break point on line 83 of DoThread.pas, this will allow you to see the event when it gets triggered.

Now press the "Manual Event" button on the application.  You will find the DoThread.pas UID is one less i.e. the event thread for some reason did not process the oldest event.  After this goes through you will see the "Manual Event" UID come through.

**Comment by bdwilton** [comment]
Just played around a bit more, if I change the "Manual Event" to call DoThread.Wake which is the same as triggering an event (in a sense), gets everything running again.

**Comment by LaKraven** [comment]
You should really use the `OnTick` procedure (it's `protected` and defined within `TLKThread`, from which `TLKEventThread` descends, and you `override` it).

Set a `TickRateLimit` equal to your desired framerate (for this demo, I'd recommend a value of 30)

Request the colors for the PaintBox on each `Tick`, and update them via the Event.

It is never a good idea to trigger UI updates directly from an Event... at least, not until I'm able to commit the update I'm working on right now specifically for UI-thread-bound Event Listeners!

**Comment by bdwilton** [comment]
Yes all fair enough, was just trying to demo the issue I struck with short schedule times and also a style of application that isn't overly UI intensive.  If using just the event engine aspect having single event triggers are a real possibility.  In my main app it doesn't seem to strike the same issue with larger than 0.3 delays.
I had glossed over the Tick procedure in TLKThread (even though I have read your precision thread bits, but sometime back) will have to give that a test as well with a more UI intensive app.

**Comment by LaKraven** [comment]
*Note: I've mistakenly said `OnTick` when I actually just mean `Tick`*

The way I'm using the Event Engine internally is to use Events to update internal member data for the Event Threads, and to trigger internal operations on those Threads.

Problem I've had (same as your initial problem here) is that trying to drive UI updates (any Synchronization from an Event Thread) disrupts the progression of the Event Threads.

- One thing you will notice is that if you override `TLKEventContainer.GetInitialThreadState` (`TLKEventContainer` is the immediate parent class of `TLKEventThread`) and return `tsRunning`, this will make the `Tick` override fire.
- You will also need to override `TLKEventContainer.GetPauseOnNoEvent` and return `False`, which will stop the Thread from going idle between Events.

Note that just doing the latter of those two things will make your test project work pretty well... but I still don't recommend doing a Synchronized call from an Event callback method, and letting the UI render at its own rate via an `Tick` override.

I have begun drafting a better use-case document for this kind of thing, and it'll be committed into the Documentation folder soon as possible (next week, as I'm on holiday after today).

**Comment by LaKraven** [comment]
Yup... I can make your demo work as-is with only one change!

In DoThread.pas, for class `TDoThread`, add the following function to the `protected` section:

`function GetPauseOnNoEvent: Boolean; override;`

Now in the implementation add the following:

```pascal
function TDoThread.GetPauseOnNoEvent: Boolean;
begin
  Result := False;
end;
```
Your program will now run indefinitely without freezing up.

*Don't forget to grab the latest updated version of the LKSL as I've also fixed the AV on shutdown!*

**Comment by LaKraven** [comment]
You'll want to switch to the "GenericsRedux" branch if you haven't already done so (this is where I've made the improvements and bug fixes, as I'm redeveloping the Generics to work properly with FreePascal as well as Delphi)

**Comment by bdwilton** [comment]
Nice, I'll give all of it a try.  Have a good holiday.

**Comment by bdwilton** [comment]
Yes the TLKEventContainer override works fine function GetPauseOnNoEvent: Boolean; override;

**Comment by LaKraven** [comment]
Hi @bdwilton 
I'll close the issue, but I will comment again when I've detached a compact demonstration of how I do rendering from a much more complicated demo project (drafting articles both for the blog, the LKSL docs and the docs for my soon-to-be-open-source game engine, AGE)
